woocommerce/woocommerce-android
# M4: Add simple products

This PR is the result of a successful trial. It adds the **Add simple product** functionality (currently behind the M4 feature-flag).

It's already been reviewed but I'd like to get a final quick look before we merge it to `develop`.

**To test:**

Try the new add-product workflow:
1. Go to Products
2. Tap on the new FAB 
3. Notice a bottom sheet with product types appears
4. Select the Physical product
5. Notice a product detail appears
6. Fill out some of the properties + add images
8. Tap on Publish button
9. Notice a new product is added (verify by refreshing the list in the Products tab)
10. Tap on the FAB again
11. Notice that the bottom sheet is not shown anymore  (by design) and the physical product type is selected automatically

**The original description:**

## RESOURCES
[The Figma file](https://www.figma.com/file/QAkdm5HZdLlG0b8vUkv1pX/Products---Add-%26-Edit-Products?node-id=460%3A1760&viewport=-1704%2C1577%2C0.5334523320198059) and [the initial issue](https://github.com/woocommerce/woocommerce-android/issues/2707)

## CHANGES / DESCRIPTION

This PR will contain the following:
- added tests for the **add product** view model flow
- added tests for the product list 
- refactored `AppPrefs` (a little bit!) to allow being injected in different places - **helps going forward with tests** 👍 

Starting from the last bit, refactored the way `AppPrefs` is used a bit by having it injected in different view models and also in the main `WooCommerce` application class. Basically, to achieve this, I added the following in the `AppConfigModule`:
```
@Provides
fun providesAppPrefs(appContext: Context): AppPrefs {
    AppPrefs.init(appContext)
    return AppPrefs
}
```
This means that it can now be injected 🎉 

I then updated the `WooCommerce` to put things in the correct order and use the injected preferences class. As far as I am aware, this should be fine, even calling `AppPrefs` from different places and directly (aka, not injected) as the `init` already happened and the context is the same from the start. 

I double checked the app running and checked the logs for analytics and such, and it all seems to be working fine...but **it needs proper testing to make sure nothing broke** 🙏 

Now for the new tests!

Added a new test scenario named `ProductDetailViewModel_AddFlowTest` (not a great naming, but it's descriptive) that will handle the **add product** test flow. It's injected with the correct `navArgs` and uses a mock `AppPrefs` in order to get the correct selected default `type` of product 👍 

I also added a test on the `ProductListViewModelTest` for the check of the **add button** visibility given the different loading scenarios.

As always, let me know your thoughts 🙏 


**EDIT**
Also:
- added the `PUBLISH` to `UPDATE` logic in order to display the top nav bar navigation and correct loading/success dialogs, depending on the flow

## SCREENSHOTS
All the tests
-
<img width="500" alt="Screenshot 2020-08-28 at 00 15 28" src="https://user-images.githubusercontent.com/7058393/91503905-98ea9a80-e8c3-11ea-9afb-9274e1107fc0.png">

Specific ones modified
-

<img width="300" alt="Screenshot 2020-08-28 at 12 44 13" src="https://user-images.githubusercontent.com/7058393/91557393-4f826580-e92c-11ea-90d5-a24b4d6c1248.png">
<img width="300" alt="Screenshot 2020-08-28 at 00 11 57" src="https://user-images.githubusercontent.com/7058393/91503724-27aae780-e8c3-11ea-879b-6014ef010699.png">
<img width="300" alt="Screenshot 2020-08-28 at 00 12 09" src="https://user-images.githubusercontent.com/7058393/91503725-28437e00-e8c3-11ea-9005-d760f769fb47.png">

Add Product flow to Update flow
-
<image src="https://user-images.githubusercontent.com/7058393/91557005-96bc2680-e92b-11ea-8c66-37e0137ad014.gif" width="350"/>
